File: MercuryAPI-1.31.3.36/c/src/samples/untraceable.c
/**
* Sample program that to demonstrate the usage of Gen2v2 untraceable.
* @file untraceable.c
*/

#include <tm_reader.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <inttypes.h>

#if WIN32
#define snprintf sprintf_s
#endif 

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif

#define usage() {errx(1, "Please provide valid reader URL, such as: reader-uri [--ant n]\n"\
                         "reader-uri : e.g., 'tmr:///COM1' or 'tmr:///dev/ttyS0/' or 'tmr://readerIP'\n"\
                         "[--ant n] : e.g., '--ant 1'\n"\
                         "Example: 'tmr:///com4' or 'tmr:///com4 --ant 1,2' \n");}

void errx(int exitval, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);

	exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
	if (TMR_SUCCESS != ret)
	{
		errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
	}
}

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
									 uint32_t timeout, void *cookie)
{
	FILE *out = cookie;
	uint32_t i;

	fprintf(out, "%s", tx ? "Sending: " : "Received:");
	for (i = 0; i < dataLen; i++)
	{
		if (i > 0 && (i & 15) == 0)
		{
			fprintf(out, "\n         ");
		}
		fprintf(out, " %02x", data[i]);
	}
	fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
	FILE *out = cookie;

	fprintf(out, "%s", tx ? "Sending: " : "Received:");
	fprintf(out, "%s\n", data);
}

void parseAntennaList(uint8_t *antenna, uint8_t *antennaCount, char *args)
{
	char *token = NULL;
	char *str = ",";
	uint8_t i = 0x00;
	int scans;

	/* get the first token */
	if (NULL == args)
	{
		fprintf(stdout, "Missing argument\n");
		usage();
	}

	token = strtok(args, str);
	if (NULL == token)
	{
		fprintf(stdout, "Missing argument after %s\n", args);
		usage();
	}

	while(NULL != token)
	{
		scans = sscanf(token, "%"SCNu8, &antenna[i]);
		if (1 != scans)
		{
			fprintf(stdout, "Can't parse '%s' as an 8-bit unsigned integer value\n", token);
			usage();
		}
		i++;
		token = strtok(NULL, str);
	}
	*antennaCount = i;
}

void ReadTags(TMR_Reader* rp)
{
	TMR_Status ret;
	ret = TMR_read(rp, 500, NULL);
	if (TMR_ERROR_TAG_ID_BUFFER_FULL == ret)
	{
		/* In case of TAG ID Buffer Full, extract the tags present
		* in buffer.
		*/
		fprintf(stdout, "reading tags:%s\n", TMR_strerr(rp, ret));
	}
	else
	{
		checkerr(rp, ret, 1, "reading tags");
	}

	while (TMR_SUCCESS == TMR_hasMoreTags(rp))
	{
		TMR_TagReadData trd;
		char epcStr[128];
		char timeStr[128];
        uint8_t dataBuf[256];

        ret = TMR_TRD_init_data(&trd, sizeof(dataBuf)/sizeof(uint8_t), dataBuf);
        checkerr(rp, ret, 1, "creating tag read data");

		ret = TMR_getNextTag(rp, &trd);
		checkerr(rp, ret, 1, "fetching tag");

		TMR_bytesToHex(trd.tag.epc, trd.tag.epcByteCount, epcStr);

#ifdef WIN32
		{
			FILETIME ft, utc;
			SYSTEMTIME st;
			char* timeEnd;
			char* end;

			utc.dwHighDateTime = trd.timestampHigh;
			utc.dwLowDateTime = trd.timestampLow;

			FileTimeToLocalFileTime( &utc, &ft );
			FileTimeToSystemTime( &ft, &st );
			timeEnd = timeStr + sizeof(timeStr)/sizeof(timeStr[0]);
			end = timeStr;
			end += sprintf(end, "%d-%d-%d", st.wYear,st.wMonth,st.wDay);
			end += sprintf(end, "T%d:%d:%d.%03d", st.wHour,st.wMinute,st.wSecond, st.wMilliseconds);
		}
#else
		{
			uint8_t shift;
			uint64_t timestamp;
			time_t seconds;
			int micros;
			char* timeEnd;
			char* end;

			shift = 32;
			timestamp = ((uint64_t)trd.timestampHigh<<shift) | trd.timestampLow;
			seconds = timestamp / 1000;
			micros = (timestamp % 1000) * 1000;

			/*
			* Timestamp already includes millisecond part of dspMicros,
			* so subtract this out before adding in dspMicros again
			*/
			micros -= trd.dspMicros / 1000;
			micros += trd.dspMicros;

			timeEnd = timeStr + sizeof(timeStr)/sizeof(timeStr[0]);
			end = timeStr;
			end += strftime(end, timeEnd-end, "%FT%H:%M:%S", localtime(&seconds));
			end += snprintf(end, timeEnd-end, ".%06d", micros);
			end += strftime(end, timeEnd-end, "%z", localtime(&seconds));
		}
#endif

        printf("EPC:%s ant:%d count:%d Time:%s\n", epcStr, trd.antenna, trd.readCount, timeStr);
        if(0 < trd.data.len)
        {
          char dataStr[255];
          TMR_bytesToHex(trd.data.list, trd.data.len, dataStr);
          printf("Raw data(%d): %s\n", trd.data.len, dataStr);
        }
	}
}

/* performEmbeddedTagOperation() performs following functions:
 * a) Set the tagOp to the read plan.
 * b) Set the filter to the read plan
 * c) set the read plan to the reader.
 * d) perform sync read.
 * e) process the data if it is there.
 * 
 * @param rp    : pointer to the reader object.
 * @param plan  : pointer to the read plan object
 * @param tagOp : pointer to the tag operation object
 * @param filter: pointer to the filter object 
 */
void performEmbeddedTagOperation(TMR_Reader *reader, TMR_ReadPlan *plan, TMR_TagOp *tagOp, TMR_TagFilter *filter)
{
  TMR_Status ret;

  // Set tagOp to read plan 
  ret = TMR_RP_set_tagop(plan, tagOp);
  checkerr(reader, ret, 1, "setting tagop");

  // Set filter to read plan
  ret = TMR_RP_set_filter(plan, filter);
  checkerr(reader, ret, 1, "setting filter");

  // Commit read plan 
  ret = TMR_paramSet(reader, TMR_PARAM_READ_PLAN, plan);
  checkerr(reader, ret, 1, "setting read plan");

  ReadTags(reader);
}

int main(int argc, char *argv[])
{
	TMR_Reader r, *rp;
	TMR_Status ret;
	TMR_ReadPlan plan;
	TMR_Region region;
	uint8_t *antennaList = NULL;
	uint8_t buffer[20];
	uint8_t i;
	uint8_t antennaCount = 0x0;
        TMR_GEN2_Session session = TMR_GEN2_SESSION_S0;
#if USE_TRANSPORT_LISTENER
	TMR_TransportListenerBlock tb;
#endif

	if (argc < 2)
	{
		usage(); 
	}

	for (i = 2; i < argc; i+=2)
	{
		if(0x00 == strcmp("--ant", argv[i]))
		{
			if (NULL != antennaList)
			{
				fprintf(stdout, "Duplicate argument: --ant specified more than once\n");
				usage();
			}
			parseAntennaList(buffer, &antennaCount, argv[i+1]);
			antennaList = buffer;
		}
		else
		{
			fprintf(stdout, "Argument %s is not recognized\n", argv[i]);
			usage();
		}
	}

	rp = &r;
	ret = TMR_create(rp, argv[1]);
	checkerr(rp, ret, 1, "creating reader");

#if USE_TRANSPORT_LISTENER

	if (TMR_READER_TYPE_SERIAL == rp->readerType)
	{
		tb.listener = serialPrinter;
	}
	else
	{
		tb.listener = stringPrinter;
	}
	tb.cookie = stdout;

	TMR_addTransportListener(rp, &tb);
#endif

	ret = TMR_connect(rp);
	checkerr(rp, ret, 1, "connecting reader");

	region = TMR_REGION_NONE;
	ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
	checkerr(rp, ret, 1, "getting region");

	if (TMR_REGION_NONE == region)
	{
		TMR_RegionList regions;
		TMR_Region _regionStore[32];
		regions.list = _regionStore;
		regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
		regions.len = 0;

		ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
		checkerr(rp, ret, __LINE__, "getting supported regions");

		if (regions.len < 1)
		{
			checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't supportany regions");
		}
		region = regions.list[0];
		ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
		checkerr(rp, ret, 1, "setting region");  
	}
  /**
   * Checking the software version of the sargas.
   * The antenna detection is supported on sargas from software version of 5.3.x.x.
   * If the Sargas software version is 5.1.x.x then antenna detection is not supported.
   * User has to pass the antenna as arguments.
   */
    {
      ret = isAntDetectEnabled(rp, antennaList);
      if(TMR_ERROR_UNSUPPORTED == ret)
      {
        fprintf(stdout, "Reader doesn't support antenna detection. Please provide antenna list.\n");
        usage();
      }
      else
      {
        checkerr(rp, ret, 1, "Getting Antenna Detection Flag Status");
      }
    }

	/**
	* for antenna configuration we need two parameters
	* 1. antennaCount : specifies the no of antennas should
	*    be included in the read plan, out of the provided antenna list.
	* 2. antennaList  : specifies  a list of antennas for the read plan.
	**/ 

	// initialize the read plan 
	ret = TMR_RP_init_simple(&plan, antennaCount, antennaList, TMR_TAG_PROTOCOL_GEN2, 1000);
	checkerr(rp, ret, 1, "initializing the  read plan");

	/* Commit read plan */
	ret = TMR_paramSet(rp, TMR_PARAM_READ_PLAN, &plan);
	checkerr(rp, ret, 1, "setting read plan");

        ret = TMR_paramSet(rp, TMR_PARAM_GEN2_SESSION, &session);
        checkerr(rp, ret, 1, "setting the session to S0");
	// Read tag epc before performing untraceable action
	ReadTags(rp);

	{
		TMR_TagOp newtagop;
		//TMR_TagFilter selectFilter;
		TMR_TagOp_GEN2_NXP_Tam1Authentication auth;
		TMR_uint8List key;
		int EpcLength = 4; //words
		bool SendRawData = false;
        bool enableEmbeddedTagop = false;
		TMR_TagOp_GEN2_NXP_Untraceable untraceable;
		//TMR_GEN2_Password accessPassword = 0x00000000;
		//uint8_t mask[128];
		uint8_t key0[] = { 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF };
		//uint8_t key1[] = { 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88};

		/* mask[0] = 0xE2;
		mask[1] = 0xE2;
		mask[2] = 0xC0;
		mask[3] = 0x6F;
		TMR_TF_init_gen2_select(&selectFilter, false, TMR_GEN2_BANK_EPC, 32, 16, mask);*/

		// Untraceable with TAM1 with Key0
        key.list = key0;
        key.max = key.len = sizeof(key0) / sizeof(key0[0]);

		ret = TMR_TagOp_init_GEN2_NXP_AES_Tam1authentication(&auth, KEY0, &key,SendRawData);
		checkerr(rp, ret, 1, "initializing Tam1 authentication");

		untraceable.auth.tam1Auth = auth;
		untraceable.auth.authType = UNTRACEABLE_WITH_AUTHENTICATION;

		ret = TMR_TagOp_init_GEN2_NXP_AES_Untraceable(&newtagop,EPC_HIDE,EpcLength,HIDE_NONE,SHOW,NORMAL,&untraceable);
		checkerr(rp, ret, 1, "initializing Untraceable");

		ret = TMR_executeTagOp(rp, &newtagop, NULL, NULL);
		checkerr(rp, ret, 1, "executing Untraceable tagop");

		ReadTags(rp);

        // Enable embedded tagOp operation by making enableEmbeddedTagop = true 
        if(enableEmbeddedTagop)
          performEmbeddedTagOperation(rp, &plan, &newtagop, NULL);

		//Uncomment this to enable untraceable with TAM1 with Key1
		/*key.list = key1;
		key.max = key.len = sizeof(key1) / sizeof(key1[0]);

		EpcLength = 6; //words
		ret = TMR_TagOp_init_GEN2_NXP_AES_Tam1authentication(&auth, KEY1, &key, SendRawData);
		checkerr(rp, ret, 1, "initializing Tam1 authentication");

		untraceable.auth.tam1Auth = auth;
		untraceable.auth.authType = UNTRACEABLE_WITH_AUTHENTICATION;

		ret = TMR_TagOp_init_GEN2_NXP_AES_Untraceable(&newtagop,EPC_HIDE,EpcLength,HIDE_NONE,SHOW,NORMAL,&untraceable);
		checkerr(rp, ret, 1, "initializing Untraceable");

		ret = TMR_executeTagOp(rp, &newtagop, NULL, NULL);
		checkerr(rp, ret, 1, "executing Untraceable tagop");

		ReadTags(rp);

        // Enable embedded tagOp operation by making enableEmbeddedTagop = true 
        if(enableEmbeddedTagop)
          performEmbeddedTagOperation(rp, &plan, &newtagop, NULL);*/

		//Uncomment this to enable untraceable with Access
		/*EpcLength = 3; //words
		untraceable.auth.authType = UNTRACEABLE_WITH_ACCESS;
		untraceable.auth.accessPassword = accessPassword;

		ret = TMR_TagOp_init_GEN2_NXP_AES_Untraceable(&newtagop,EPC_HIDE,EpcLength,HIDE_NONE,SHOW,NORMAL, &untraceable);
		checkerr(rp, ret, 1, "initializing Untraceable");

		ret = TMR_executeTagOp(rp, &newtagop, NULL, NULL);
		checkerr(rp, ret, 1, "executing Untraceable tagop");

		ReadTags(rp);

        // Enable embedded tagOp operation by making enableEmbeddedTagop = true 
        if(enableEmbeddedTagop)
          performEmbeddedTagOperation(rp, &plan, &newtagop, NULL);*/
	}

	TMR_destroy(rp);
	return 0;
}

File: MercuryAPI-1.31.3.36/java/samples_nb/src/samples/Untraceable.java
/*
 * Sample program that demonstrates the usage of
 * Gen2v2 Untraceable
 */
package samples;

import com.thingmagic.Gen2;
import com.thingmagic.Reader;
import com.thingmagic.ReaderException;
import com.thingmagic.ReaderUtil;
import com.thingmagic.SimpleReadPlan;
import com.thingmagic.TMConstants;
import com.thingmagic.TagFilter;
import com.thingmagic.TagOp;
import com.thingmagic.TagProtocol;
import com.thingmagic.TagReadData;
import com.thingmagic.TransportListener;

public class Untraceable
{
    private static Reader r = null;
    private static int[] antennaList = null;
    static BlockPermalock.SerialPrinter serialPrinter;
    static BlockPermalock.StringPrinter stringPrinter;
    static TransportListener currentListener;
    static boolean sendRawData = false;

    static void usage()
    {
        System.out.printf("Usage: Please provide valid arguments, such as:\n"
                + "Untraceable [-v] [reader-uri] [--ant n[,n...]] \n" +
                  "-v  Verbose: Turn on transport listener\n" +
                  "reader-uri  Reader URI: e.g., \"tmr:///COM1\", \"tmr://astra-2100d3\"\n"
                + "--ant  Antenna List: e.g., \"--ant 1\", \"--ant 1,2\"\n"
                + "e.g: tmr:///com1 --ant 1,2 ; tmr://10.11.115.32 --ant 1,2\n ");
        System.exit(1);
    }

    public static void setTrace(Reader r, String args[])
    {
        if (args[0].toLowerCase().equals("on"))
        {
            r.addTransportListener(Reader.simpleTransportListener);
            currentListener = Reader.simpleTransportListener;
        } else if (currentListener != null)
        {
            r.removeTransportListener(Reader.simpleTransportListener);
        }
    }
    
    static  int[] parseAntennaList(String[] args,int argPosition)
    {
        int[] antennaList = null;
        try
        {
            String argument = args[argPosition + 1];
            String[] antennas = argument.split(",");
            int i = 0;
            antennaList = new int[antennas.length];
            for (String ant : antennas)
            {
                antennaList[i] = Integer.parseInt(ant);
                i++;
            }
        }
        catch (IndexOutOfBoundsException ex)
        {
            System.out.println("Missing argument after " + args[argPosition]);
            usage();
        }
        catch (Exception ex)
        {
            System.out.println("Invalid argument at position " + (argPosition + 1) + ". " + ex.getMessage());
            usage();
        }
        return antennaList;
    }
    
    public static void main(String argv[])
    {
        // Program setup   
        TagFilter target = null;
        int nextarg = 0;
        boolean trace = false;
        // To enable select filter, set enableFilter flag to true
        boolean enableFilter = false;
        int epcLen;
        if (argv.length < 1 || (argv.length == 1 && argv[nextarg].equalsIgnoreCase("-v")))
        {
            usage();
        }

        if (argv[nextarg].equals("-v"))
        {
            trace = true;
            nextarg++;
        }

        // Create Reader object, connecting to physical device
        try
        {

            String readerURI = argv[nextarg];
            nextarg++;

            for (; nextarg < argv.length; nextarg++)
            {
                String arg = argv[nextarg];
                if (arg.equalsIgnoreCase("--ant"))
                {
                    if (antennaList != null)
                    {
                        System.out.println("Duplicate argument: --ant specified more than once");
                        usage();
                    }
                    antennaList = parseAntennaList(argv, nextarg);
                    nextarg++;
                } else
                {
                    System.out.println("Argument " + argv[nextarg] + " is not recognised");
                    usage();
                }
            }

            r = Reader.create(readerURI);
            if (trace)
            {
                setTrace(r, new String[]
                {
                    "on"
                });
            }
            r.connect();
            
            if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet("/reader/region/id")) 
            {
                Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                if (supportedRegions.length < 1) 
                {
                    throw new Exception("Reader doesn't support any regions");
                } 
                else 
                {
                    r.paramSet("/reader/region/id", supportedRegions[0]);
                }
            }

            if (r.isAntDetectEnabled(antennaList))
            {
                System.out.println("Module doesn't has antenna detection support, please provide antenna list");
                r.destroy();
                usage();
            }

            r.paramSet("/reader/tagop/antenna", antennaList[0]);
            SimpleReadPlan plan = new SimpleReadPlan(antennaList, TagProtocol.GEN2, null, null, 1000);
            r.paramSet(TMConstants.TMR_PARAM_READ_PLAN, plan);

            /* set the session to S0 */
            System.out.println("setting the session to S0");
            r.paramSet(TMConstants.TMR_PARAM_GEN2_SESSION, Gen2.Session.S0);

            //Untrace TAM1
            byte[]  key1  =  new byte[]{(byte) 0x11, (byte) 0x22, (byte) 0x33, (byte) 0x44, (byte) 0x55, (byte) 0x66, (byte) 0x77, (byte) 0x88, (byte) 0x11, (byte) 0x22, (byte) 0x33, (byte) 0x44, (byte) 0x55, (byte) 0x66, (byte) 0x77, (byte) 0x88};
            byte[]  key0  =  new byte[]{(byte) 0x01, (byte) 0x23, (byte) 0x45, (byte) 0x67, (byte) 0x89, (byte) 0xAB, (byte) 0xCD, (byte) 0xEF, (byte) 0x01, (byte) 0x23, (byte) 0x45, (byte) 0x67, (byte) 0x89, (byte) 0xAB, (byte) 0xCD, (byte) 0xEF};
            Gen2.NXP.AES.Tam1Authentication tam1 ;

            if(enableFilter)
            {
                target = new Gen2.Select(false, Gen2.Bank.EPC, 32, 16, new byte[] {(byte)0x11,(byte)0x22});
            }

            readTag(r);

            //show 4 epc words with TAM1 + Key0
            epcLen = 4;
            tam1  = new Gen2.NXP.AES.Tam1Authentication(Gen2.NXP.AES.KeyId.KEY0, key0, sendRawData);
            Gen2.Untraceable untraceTam1 = new Gen2.NXP.AES.Untraceable(
                    Gen2.Untraceable.EPC.HIDE,epcLen, 
                    Gen2.Untraceable.TID.HIDE_SOME, 
                    Gen2.Untraceable.User.SHOW,
                    Gen2.Untraceable.Range.NORMAL, 
                    tam1);

            System.out.println(untraceTam1.toString());
            r.executeTagOp(untraceTam1, target);
            readTag(r);

            //show 6 epc words with TAM1 + Key1
            /*epcLen = 6;
            tam1 = new Gen2.NXP.AES.Tam1Authentication(Gen2.NXP.AES.KeyId.KEY1, key1, sendRawData);
            Gen2.Untraceable untraceTam1Key1 = new Gen2.NXP.AES.Untraceable(
                    Gen2.Untraceable.EPC.HIDE, epcLen,
                    Gen2.Untraceable.TID.HIDE_SOME,
                    Gen2.Untraceable.User.SHOW,
                    Gen2.Untraceable.Range.NORMAL,
                    tam1);

            System.out.println(untraceTam1Key1.toString());
            r.executeTagOp(untraceTam1Key1, target);
            readTag(r);
            */

            //show 6 epc words with UntraceAcess

            /*Gen2.Password pwd = new Gen2.Password(0x00000000);
            Gen2.Untraceable untraceAcess = new Gen2.NXP.AES.Untraceable(Gen2.Untraceable.EPC.HIDE,6, Gen2.Untraceable.TID.HIDE_NONE, Gen2.Untraceable.User.SHOW, Gen2.Untraceable.Range.NORMAL, pwd.getValue());
            System.out.println(untraceAcess.toString());
            r.executeTagOp(untraceAcess, target);
            readTag(r);
            */

            /* Embedded tag operations */
            {

                //show 6 epc words with TAM1 + Key0
                /*epcLen = 6;
                tam1 = new Gen2.NXP.AES.Tam1Authentication(Gen2.NXP.AES.KeyId.KEY0, key0, sendRawData);
                Gen2.Untraceable embeddeduntraceTam1Key0 = new Gen2.NXP.AES.Untraceable(
                        Gen2.Untraceable.EPC.HIDE, epcLen,
                        Gen2.Untraceable.TID.HIDE_SOME,
                        Gen2.Untraceable.User.SHOW,
                        Gen2.Untraceable.Range.NORMAL,
                        tam1);

                System.out.println("Embedded Tag operation of " + embeddeduntraceTam1Key0.toString());
                performEmbeddedOperation(target, embeddeduntraceTam1Key0);
                readTag(r);
                */

                //show 6 epc words with TAM1 + Key1
                /*epcLen = 6;
                tam1 = new Gen2.NXP.AES.Tam1Authentication(Gen2.NXP.AES.KeyId.KEY1, key1, sendRawData);
                Gen2.Untraceable embeddeduntraceTam1Key1 = new Gen2.NXP.AES.Untraceable(
                        Gen2.Untraceable.EPC.HIDE, epcLen,
                        Gen2.Untraceable.TID.HIDE_SOME,
                        Gen2.Untraceable.User.SHOW,
                        Gen2.Untraceable.Range.NORMAL,
                        tam1);

                System.out.println("Embedded Tag operation of " + embeddeduntraceTam1Key1.toString());
                performEmbeddedOperation(target, embeddeduntraceTam1Key1);
                readTag(r);
                */

                //show 6 epc words with UntraceAcess
                /*Gen2.Password password = new Gen2.Password(0x00000000);
                Gen2.Untraceable embeddeduntraceAcess = new Gen2.NXP.AES.Untraceable(Gen2.Untraceable.EPC.HIDE, 6, Gen2.Untraceable.TID.HIDE_NONE, Gen2.Untraceable.User.SHOW, Gen2.Untraceable.Range.NORMAL, password.getValue());
                System.out.println("Embedded Tag operation of " + embeddeduntraceAcess.toString());
                performEmbeddedOperation(target, embeddeduntraceAcess);
                readTag(r);
                */

            }

        }
        catch(Exception e)
        {
            e.printStackTrace();
        }
    }
    
    public static void readTag(Reader r) throws ReaderException
    {
        TagReadData[] tagReads = r.read(500);
        for (TagReadData tr : tagReads)
        {
            System.out.println(tr.getTag());
        }
    }
    
    public static void performEmbeddedOperation(TagFilter filter, TagOp op) throws Exception
    {
        TagReadData[] tagReads = null;
        byte[] response = null;
        SimpleReadPlan plan = new SimpleReadPlan(antennaList, TagProtocol.GEN2, filter, op, 1000);
        r.paramSet("/reader/read/plan", plan);
        tagReads = r.read(1000);
        for (TagReadData tr : tagReads)
        {
            response = tr.getData();
        }
    }
}
